File: backend/app/api/routes/evaluation.py
from fastapi import APIRouter, HTTPException
from backend.app.schemas.student import MCQProblem, MCQSubmission
from backend.app.services.knowledge_base import KnowledgeBase
import random

router = APIRouter(prefix="/evaluation", tags=["evaluation"])

@router.get("/problems/{topic_id}/{concept_id}", response_model=MCQProblem)
def get_problem_for_concept(topic_id: str, concept_id: str):
    """Fetches a random MCQ problem for a specific concept."""
    try:
        kb = KnowledgeBase(topic_id)
    except FileNotFoundError:
        raise HTTPException(status_code=404, detail="Topic not found")

    # Find all problems that test this concept
    matching_problems = [p for p in kb.problems.get("items", []) if concept_id in p.get("concept_ids", [])]
    
    if not matching_problems:
        raise HTTPException(status_code=404, detail=f"No problems found for concept {concept_id}")

    # Pick a random problem
    selected_problem = random.choice(matching_problems)
    return selected_problem


@router.post("/evaluate")
def evaluate_submission(submission: MCQSubmission, topic_id: str):
    """Evaluates a student's answer deterministically."""
    try:
        kb = KnowledgeBase(topic_id)
    except FileNotFoundError:
        raise HTTPException(status_code=404, detail="Topic not found")

    problem = next((p for p in kb.problems.get("items", []) if p["id"] == submission.problem_id), None)
    if not problem:
        raise HTTPException(status_code=404, detail="Problem not found")

    is_correct = (submission.selected_option == problem["correct_option"])

    if is_correct:
        return {
            "mastery_achieved": True,
            "replan_required": False,
            "feedback": "Correct! You have mastered this concept."
        }
    else:
        return {
            "mastery_achieved": False,
            "replan_required": True,
            "feedback": f"Incorrect. The correct answer was {problem['correct_option']}."
        }

File: backend-node/src/routes/evaluation.js
import { Router } from "express";
import { KnowledgeBase } from "../services/knowledgeBase.js";
import { buildQuestionGenerationContext } from "../services/retrievalService.js";

const router = Router();

function publicProblem(problem) {
  return {
    id: problem.id,
    title: problem.title,
    concept_ids: problem.concept_ids || [],
    difficulty: problem.difficulty,
    question_text: problem.question_text,
    options: problem.options || {},
    expected_error_tags: problem.expected_error_tags || [],
  };
}

router.get("/evaluation/problems/:topicId/:conceptId", async (req, res, next) => {
  try {
    const { topicId, conceptId } = req.params;
    const studentId = req.query.student_id || null;
    let kb;
    try {
      kb = new KnowledgeBase(topicId);
    } catch (err) {
      return res.status(404).json({ detail: "Topic not found" });
    }

    const allProblems = kb.problems?.items || [];
    const matchingProblems = allProblems.filter((p) =>
      (p.concept_ids || []).includes(conceptId)
    );

    if (matchingProblems.length === 0) {
      return res.status(404).json({ detail: `No problems found for concept ${conceptId}` });
    }

    let selectedProblem = matchingProblems[Math.floor(Math.random() * matchingProblems.length)];

    if (studentId) {
      const retrieval = await buildQuestionGenerationContext(studentId, [conceptId]);
      const preferredQuestionIds = new Set(
        (retrieval.graph_context || []).map((row) => row.question_id).filter(Boolean)
      );
      if (preferredQuestionIds.size > 0) {
        const preferred = matchingProblems.filter((p) => preferredQuestionIds.has(p.id));
        if (preferred.length > 0) {
          selectedProblem = preferred[Math.floor(Math.random() * preferred.length)];
        }
      }
    }

    res.json(publicProblem(selectedProblem));
  } catch (err) {
    next(err);
  }
});

router.post("/evaluation/evaluate", (req, res, next) => {
  try {
    const topicId = req.query.topic_id;
    if (!topicId) {
      return res.status(400).json({ detail: "topic_id query param is required" });
    }

    const { problem_id, selected_option } = req.body;
    let kb;
    try {
      kb = new KnowledgeBase(topicId);
    } catch (err) {
      return res.status(404).json({ detail: "Topic not found" });
    }

    const problem = (kb.problems?.items || []).find((p) => p.id === problem_id);
    if (!problem) {
      return res.status(404).json({ detail: "Problem not found" });
    }

    const isCorrect = selected_option === problem.correct_option;

    if (isCorrect) {
      res.json({
        mastery_achieved: true,
        replan_required: false,
        feedback: "Correct. This checkpoint suggests the learner is moving in the right direction.",
      });
    } else {
      res.json({
        mastery_achieved: false,
        replan_required: true,
        feedback: "Not quite. This miss will be used to adjust the next plan rather than relying on manual form input.",
      });
    }
  } catch (err) {
    next(err);
  }
});

export default router;
